### PR TITLE
#162727: 'Debug:Run to cursor' command is now available in inactive d…

### PR DESCRIPTION
**Pull request by Viijay-Kr**
Addresses the feature requested in https://github.com/microsoft/vscode/issues/162727

This PR took the strategy mentioned in this [comment](https://github.com/microsoft/vscode/issues/162727#issuecomment-1268634341) 

## Refactorings
1. The `debuggerService.runTo` is refactored to support inactive debugger state
2. Two closures have been added to handle add and remove temp breakpoints.
3. Remove breakpoints is attached to the listeners in the both the cases (Stopped and Inactive blocks)

### CODE REVIEW

**Comment by Viijay-Kr** [comment]
@microsoft-github-policy-service agree

File: src/vs/workbench/contrib/debug/browser/debugEditorActions.ts [hunk 1]
```diff
@@ -230,7 +231,29 @@ export class RunToCursorAction extends EditorAction {
 			// otherwise set it at the precise column #102199
 			column = position.column;
 		}
-
+		if (debugService.state === State.Inactive) {
+			// add a break point to be removed later
```
Comment (roblourens) [inline src/vs/workbench/contrib/debug/browser/debugEditorActions.ts:235]: This appears to duplicate the logic in `debugService.runTo`, I'd prefer to find a way to reuse that code somehow

Comment (Viijay-Kr) [reply src/vs/workbench/contrib/debug/browser/debugEditorActions.ts:235]: @roblourens This seems a bit challenging . Run to works under the assumption that the debugger is in stop state.
If I try to add the inactive condition and start the debugger , the temporary breakpoint won't get removed.

Do you have any ideas on how to overcome this ?

Comment (roblourens) [reply src/vs/workbench/contrib/debug/browser/debugEditorActions.ts:235]: > If I try to add the inactive condition and start the debugger inside the runTo , the temporary breakpoint won't get removed.

I don't understand the problem

Comment (Viijay-Kr) [reply src/vs/workbench/contrib/debug/browser/debugEditorActions.ts:235]: Inside the `debuggerService.runTo`, If I add the logic to start the debugger when it is in `INACTIVE` state 
(preferably [here](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugService.ts#L1171) )

```ts
let focusedSession = this.getViewModel().focusedSession;
if (this.state === State.Inactive) {
	const { launch, name, getConfig } = this.getConfigurationManager().selectedConfiguration;
	const config = await getConfig();
	const configOrName = config ? Object.assign(deepClone(config), {}) : name;
	await this.startDebugging(launch, configOrName, undefined, true);
	focusedSession = this.getViewModel().focusedSession;
}
else if (this.state !== State.Stopped) {
	return;
}
if (!focusedSession) {
	return;
}
```

, the breakpoint which gets added [here](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugService.ts#L1194) (to be removed later) won't get removed . 

```ts
const oneTimeListener = threadToContinue.session.onDidChangeState(() => {
const state = focusedSession?.state;
if (state === State.Stopped || state === State.Inactive) {
if (breakpointToRemove) {
	this.removeBreakpoints(breakpointToRemove.getId());
}
oneTimeListener.dispose();
}
});
```

Isn't that the expected behaviour of the `runTo` action the breakpoint is temp and not permanent .

Comment (roblourens) [reply src/vs/workbench/contrib/debug/browser/debugEditorActions.ts:235]: > won't get removed .

Why not? I still don't understand. Seems like I am looking at two methods that have most of their logic in common and I am sure there is a way to share most of it.

Comment (Viijay-Kr) [reply src/vs/workbench/contrib/debug/browser/debugEditorActions.ts:235]: Well the reason is due to the fact that 'threadToContinue' is 'undefined' after the debugger is started.


I get your suggestion about reusing the logic. However I'm still not sure how to . 

One idea I have is to separate the logic that can be abstracted out of the 'debuggerService.runTo' namely 
1. The logic to add the temp breakPoint 
2. The logic to remove the temp breakpoint

Comment (roblourens) [reply src/vs/workbench/contrib/debug/browser/debugEditorActions.ts:235]: Makes sense to separate out some code that can be shared, or rewrite it in a way that you will have a session to listen to in either case.

**Review by roblourens — APPROVED** [review]
Thanks!

**Review by Tyriar — APPROVED** [review]

**Review by mjbvz — APPROVED** [review]

**Review by mjbvz — APPROVED** [review]

**Pull request merged**